app/upgrades/v20/upgrades.go

```diff
@@ -25,7 +46,80 @@ func CreateUpgradeHandler(
 			return vm, err
 		}
 
+		InitializeMaxProviderConsensusParam(ctx, keepers.ProviderKeeper)
+		InitializeMissingValidatorSetCaps(ctx, keepers.ProviderKeeper)
+		SetMaxValidators(ctx, *keepers.StakingKeeper)
+		InitializeLastProviderConsensusValidatorSet(ctx, keepers.ProviderKeeper, *keepers.StakingKeeper)
+
 		ctx.Logger().Info("Upgrade v20 complete")
 		return vm, nil
 	}
 }
+
+// InitializeMissingValidatorSetCaps initializes the max validators
+// parameter for existing consumers to the MaxProviderConsensusValidators parameter.
+// This is necessary to avoid those consumer chains having an excessive amount of validators.
+func InitializeMissingValidatorSetCaps(ctx sdk.Context, providerKeeper providerkeeper.Keeper) {
+	for _, chainID := range providerKeeper.GetAllRegisteredConsumerChainIDs(ctx) {
+		providerKeeper.SetValidatorSetCap(ctx, chainID, NewMaxProviderConsensusValidators)
+	}
+}
+
+// InitializeMaxProviderConsensusParam initializes the MaxProviderConsensusValidators parameter.
+// It is set to 180, which is the current number of validators participating in consensus on the Cosmos Hub.
+// This parameter will be used to govern the number of validators participating in consensus on the Cosmos Hub,
+// and takes over this role from the MaxValidators parameter in the staking module.
+func InitializeMaxProviderConsensusParam(ctx sdk.Context, providerKeeper providerkeeper.Keeper) {
+	params := providerKeeper.GetParams(ctx)
+	if params.MaxProviderConsensusValidators == 0 {
```

Comment:
Why is this if condition necessary?

Reply:
Not necessary, just double-checking. Can also remove this